### PR TITLE
Vivek shankar/fixed extra node added azure api

### PR DESCRIPTION
**Pull request by vivekshankar1**
### :nut_and_bolt: Description: What code changed, and why?

Issues
- Fixed issue where extra node was being added after every azure-api scan.
- Fixed issue where nodes were added after edit nodeIntegration for azure.
- Fixed issue where nodes were added after edit nodeIntegration for aws (implementations are different for azure and aws).

Code Changes
- For Azure we are passing the source_region as ""(empty)
- This causes the NodeSelectQuery to return a query (trimmed) with a where clause like this:
select * from nodes where source_region like '%';
- We are matching NULL as well now apart from the wildcard so that it returns entries where source_region is NULL:
select * from nodes where (source_region like '%' or source_region is NULL);
- On edit integration for azure: New nodes were being added on azure update integration because when inspect scan ran after azure api integration, it used to update the `source_id, source_region and source_account_id` to null, if any of them ware empty string. For azure the source region used to be empty , thus new nodes were added
- On edit integration for AWS: When new AWS integration was added values for source region was set as was alias, but inpec scan was setting to AWS region. We are now setting it as region instead of alias
- New Nodes on Azure API scan: As due to the above reasons the unique constraint were breaking for nodes table in nodemanager-service, duplicate entries of nodes were created which resulted in new nodes per scan

### :chains: Related Resources
https://github.com/chef/automate/issues/5270

### :+1: Definition of Done

Extra nodes should not be added after every azure API scan. Number of nodes must remain the same

### :athletic_shoe: How to Build and Test the Change

- rebuild components/nodemanager-service

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)-Not required as its a bug

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :white_check_mark: Note

For customers facing this issue, should remove their existing misbehaving **node integrations** and they should recreate it.

### CODE REVIEW

**Review by vivek-yadav — APPROVED** [review]

**Review by kalroy — CHANGES_REQUESTED** [review]
Although the issue is fixed, check the review comments.

File: components/nodemanager-service/pgdb/filtering.go [hunk 1]
```diff
@@ -206,24 +206,42 @@ func patternMatchTags(field string, arr []string, tableAbbrev string) (condition
 	return strings.Join(valueLikes, " OR "), nil
 }
 
+func Unique(a []string) []string {
+	seen := make(map[string]struct{})
```
[the remark this answers is not in the document]
Comment (vivekshankar1) [reply components/nodemanager-service/pgdb/filtering.go]: I have made the requested changes.

File: components/nodemanager-service/pgdb/filtering.go [hunk 2]
```diff
@@ -206,24 +206,42 @@ func patternMatchTags(field string, arr []string, tableAbbrev string) (condition
...
+	b := a[:0]
```
[the remark this answers is not in the document]
Comment (vivekshankar1) [reply components/nodemanager-service/pgdb/filtering.go]: I have made the requested changes.

File: components/nodemanager-service/pgdb/filtering.go [hunk 3]
```diff
@@ -206,24 +206,42 @@ func patternMatchTags(field string, arr []string, tableAbbrev string) (condition
...
 	if len(arr) == 0 {
-		condition += "''"
+		condition += "'')"
 		return condition, nil
 	}
+	arr = Unique(arr)
 	for index, item := range arr {
 		item = strings.TrimSuffix(item, "*")
 		condition += fmt.Sprintf("'%s%%'", pgutils.EscapeLiteralForPGPatternMatch(item))
+		if item == "" {
```
[the remark this answers is not in the document]
Comment (vivekshankar1) [reply components/nodemanager-service/pgdb/filtering.go]: No there is no performance concern as coalesce is more like a case. I have made the requested changes.

File: components/nodemanager-service/pgdb/filtering.go [hunk 4]
```diff
@@ -206,24 +206,42 @@ func patternMatchTags(field string, arr []string, tableAbbrev string) (condition
...
+			condition += fmt.Sprintf(" OR %s.%s is NULL", tableAbbrev, field)
+		}
 		if index < len(arr)-1 {
-			condition += fmt.Sprintf(" OR %s.%s LIKE ", tableAbbrev, field)
+			condition += ")"
```
[the remark this answers is not in the document]
Comment (vivekshankar1) [reply components/nodemanager-service/pgdb/filtering.go]: removed it as not required any more after coalesce

File: components/nodemanager-service/pgdb/filtering.go [hunk 5]
```diff
@@ -206,24 +206,42 @@ func patternMatchTags(field string, arr []string, tableAbbrev string) (condition
...
+			condition += fmt.Sprintf(" OR (%s.%s LIKE ", tableAbbrev, field)
 		}
 	}
 
+	condition += ")"
```
[the remark this answers is not in the document]
Comment (vivekshankar1) [reply components/nodemanager-service/pgdb/filtering.go]: removed it as not required any more after coalesce

**Review by vivek-yadav — APPROVED** [review]

**Review by kalroy — APPROVED** [review]

**Pull request merged**